### ISSUE TITLE
Locally include static elements in a PyTree traversing?

### ISSUE THREAD

**Issue by adonath** [jax-ml/jax#30860]
I have found `jax.tree.map` and similar functions to be useful for walking and manipulating PyTrees in general. However one limitation for this kind of "off label use"  is that static elements are always excluded. Consider the following minimal example:

```python
import jax
from dataclasses import dataclass

@jax.tree_util.register_dataclass
@dataclass
class MyStruct:
    value: jax.Array
    name: str = field(default="par-a", metadata=dict(static=True))


# Intended use case...
jax.tree.map(lambda x: x **2, MyStruct(value=jax.numpy.ones((2, 2))))

def print_everything(leaf):
    print(leaf)
    return leaf

# However this can be useful as well, but only prints the array and excludes static members...
jax.tree.map(print_everything, MyStruct(value=jax.numpy.ones((2, 2))))

# So should this exist?
jax.tree.map(print_everything, MyStruct(value=jax.numpy.ones((2, 2))), include_static=True)
```
While I understand the intended use cases, I wonder whether there is a simple way of making the above example with the printing work?

I guess the answer is likely no, because the flattening / unflattening is tied to the class and static. One would need to add context to be able to serialize and de-serialize in a configurable way. But I am happy to hear any further ideas.

**Comment by jakevdp** [comment]
I don't think this could be supported in general. `tree_map` over dynamic elements works because during flattening, the dynamic elements (i.e. `children`) are always returned as a flat list, and this can be mapped over. On the other hand, the static elements (i.e. `aux_data`) has no such constraint: the only requirement is that `aux_data` be a hashable and immutable Python object with suitable equality semantics. Given that, in general there's no way to map over static contents of a pytree unless we were to radically change the specification of tree flattening.

**Comment by adonath** [comment]
Thanks @jakevdp for the quick reply! The lack of a serialization convention for the static elements makes this idea indeed unreasonable and one would rather rely on an external solution (probably `optree` or naive Python implementation).

Feel free to close, unless you think there might be some path forward.

**Comment by adonath** [comment]
Just one note: for some cases with dataclasses and when there is no output needed from `map` one can locally use `asdict()` to convert to a plain dict, which includes the static data.

**Comment by jakevdp** [comment]
Closing because the requested feature is infeasible. Thanks!
